packages/react/src/components/FocusTrapZone/FocusTrapZone.tsx

```diff
@@ -97,284 +92,230 @@ export const FocusTrapZone: React.FunctionComponent<IFocusTrapZoneProps> & {
-        (elementContains(root.current, activeElement) || activeElement === doc.body)
-      ) {
-        if (
-          !(
-            internalState.previouslyFocusedElementOutsideTrapZone === firstBumper.current ||
-            internalState.previouslyFocusedElementOutsideTrapZone === lastBumper.current
-          )
-        ) {
-          focusAsync(internalState.previouslyFocusedElementOutsideTrapZone);
-        }
-      }
+    if (!doc) {
+      return;
     }
-  }, [doc, id, ignoreExternalFocusing, internalState]);
 
-  const forceFocusInTrap = React.useCallback(
-    (ev: FocusEvent): void => {
-      if (disabled) {
-        return;
-      }
-      if (FocusTrapZone.focusStack.length && id === FocusTrapZone.focusStack[FocusTrapZone.focusStack.length - 1]) {
-        const focusedElement = ev.target as HTMLElement;
-        if (!elementContains(root.current, focusedElement)) {
-          focus();
-          internalState.hasFocus = true; // set focus here since we stop event propagation
-          ev.preventDefault();
-          ev.stopPropagation();
-        }
-      }
-    },
-    [disabled, id, focus, internalState],
-  );
+    const activeElement = doc.activeElement as HTMLElement;
+    if (
+      !disableRestoreFocus &&
+      typeof elementToFocusOnDismiss?.focus === 'function' &&
+      // only restore focus if the current focused element is within the FTZ, or if nothing is focused
+      (elementContains(root.current, activeElement) || activeElement === doc.body)
+    ) {
+      focusElementAsync(elementToFocusOnDismiss);
+    }
+  });
 
-  const forceClickInTrap = React.useCallback(
-    (ev: MouseEvent): void => {
-      if (disabled) {
-        return;
-      }
-      if (FocusTrapZone.focusStack.length && id === FocusTrapZone.focusStack[FocusTrapZone.focusStack.length - 1]) {
-        const clickedElement = ev.target as HTMLElement;
-        if (clickedElement && !elementContains(root.current, clickedElement)) {
-          focus();
-          internalState.hasFocus = true; // set focus here since we stop event propagation
-          ev.preventDefault();
-          ev.stopPropagation();
-        }
+  /** Called in window event handlers. (useEventCallback ensures latest prop values are used.) */
+  const forceFocusOrClickInTrap = useEventCallback((ev: FocusEvent | MouseEvent): void => {
+    // be sure to use the latest values here
+    if (disabled) {
+      return;
+    }
+    if (internalState.focusStackId === FocusTrapZone.focusStack!.slice(-1)[0]) {
+      const targetElement = ev.target as HTMLElement | null;
+      if (targetElement && !elementContains(root.current, targetElement)) {
+        focusFTZ();
+        internalState.hasFocus = true; // set focus here since we stop event propagation
+        ev.preventDefault();
+        ev.stopPropagation();
       }
-    },
-    [disabled, id, focus, internalState],
-  );
-
-  const updateEventHandlers = React.useCallback((): void => {
```

Comment:
This complicated logic was trying to emulate the cleanup behavior that can now be handled automatically with effect cleanup